src/storage/storage_structure/lists/lists_update_store.cpp

```diff
@@ -88,6 +102,46 @@ void ListsUpdateStore::insertRelIfNecessary(shared_ptr<ValueVector>& srcNodeIDVe
     }
 }
 
+void ListsUpdateStore::deleteRelIfNecessary(const shared_ptr<ValueVector>& srcNodeIDVector,
+    const shared_ptr<ValueVector>& dstNodeIDVector, const shared_ptr<ValueVector>& relIDVector) {
+    auto srcNodeID =
+        srcNodeIDVector->getValue<nodeID_t>(srcNodeIDVector->state->getPositionOfCurrIdx());
+    auto dstNodeID =
+        dstNodeIDVector->getValue<nodeID_t>(dstNodeIDVector->state->getPositionOfCurrIdx());
+    auto relID = relIDVector->getValue<int64_t>(relIDVector->state->getPositionOfCurrIdx());
+    auto tupleIdx = isRelStoredInUpdateStore(relID);
+    if (tupleIdx != -1) {
+        // If the rel that we are going to delete is stored in updateStore, we simply need to delete
+        // its tupleIdx from the insertedRelsTupleIdxInFT of listUpdateStore in FWD and BWD
+        // direction.
+        for (auto direction : REL_DIRECTIONS) {
+            auto boundNodeID = direction == RelDirection::FWD ? srcNodeID : dstNodeID;
+            if (listUpdatesPerTablePerDirection[direction].contains(boundNodeID.tableID)) {
+                auto& insertedRelsTupleIdxInFT =
+                    listUpdatesPerTablePerDirection[direction]
+                        .at(boundNodeID.tableID)[StorageUtils::getListChunkIdx(boundNodeID.offset)]
+                                                [boundNodeID.offset]
+                        .insertedRelsTupleIdxInFT;
+                assert(find(insertedRelsTupleIdxInFT.begin(), insertedRelsTupleIdxInFT.end(),
+                           tupleIdx) != insertedRelsTupleIdxInFT.end());
+                insertedRelsTupleIdxInFT.erase(remove(insertedRelsTupleIdxInFT.begin(),
```

Comment:
I notice that you are not fixing your ListUpdates, ListUpdatesPerNode and ListUpdatesPerChunk indices if a rel was inserted and then deleted, so there is currently no update on the node. This is probably fine, but: (i) I would test this case explicitly; and (ii) I would also add a note somewhere, probably in ListUpdateStore that the ListUpdates can exist for a nodeOffset in ListUpdatesPerNode but be empty and give the example of adding a new rel and deleting it for a node. Similarly ListUpdatesPerChunk can exist in ListUpdatesPerChunk but be empty.

Reply:
1) Added a test about this case: DeleteAllInsertedRels.
2) I think the listUpdates and listUpdatesPerNode should be fixed prior to prepareCommit(). We can discuss this further in the next transaction PR. For now, i added a comment there.